contribute/design/1.x/csi/20190805-csi-zfspv-volume-provisioning.md

```diff
@@ -109,48 +109,105 @@ spec:
 ## Workflow
 
 ### 1. CSI create volume
-At CSI when we get a volume create request, we will create a PV object. The actual volume will be created at the Node publish time. We will have StorageClass at this point. We will create a ZfsVolume custom resource with all the volume properties that are present in the PVC and corresponding Storage Class . This custom resource will be used for creating and managing the ZFS dataset.
+At CSI, when we get a Create Volume request, it will first try to find a node where it can create the PV object. The driver will trigger the scheduler which will return a node where the PV should be created.
+In CreateVolume call, we will have the list of nodes where the ZFS pools are present and the volume should be created in anyone of the node present in the list.
+
+#### 1.1 Scheduler
+At this point the ZFS driver will have list of all nodes where ZFS pools are present. It will go through the list and pick the node where less number of volumes are provisioned.
+Lets say there are 2 nodes node1 and node2 with below pool configuration :-
+```
+node1 
+|
+|-----> pool1
+|         |
+|         |------> pvc1
+|         |------> pvc2
+|-----> pool2
+          |------> pvc3
+
+node2
+|
+|-----> pool1
+|         |
+|         |------> pvc4
+|-----> pool2
+          |------> pvc5
+	  |------> pvc6
+```
+So if application is using pool1 as shown in the below storage class, then ZFS driver will schedule it on node2 as it has one volume as compared to node1 which has 2 volumes in pool1.
```

Comment:
Will this scheduling be done as a separate CR. In other words, will it lead to creation of a CR that has the volume, pool & node details? 
_OR_
Will these info be passed along as inputs/responses & the targeted agent/process parses the payload & gets it done?

Reply:
I am trying to understand if we care about saving the evaluation of scheduling decision into a CR & let the corresponding watcher do the rest

Reply:
Above CR if we have can help us in doing a manual schedule if required?

Reply:
there is  no separate CR for scheduling, here in this case scheduling has to be part of CreateVolume call only as we have to create the PV on a node.

Reply:
this is a good start on scheduling @pawanpraka1. it is probably good idea to consider this case. 

(a) Schedule based on the number of volumes ( which is what is currently proposed )
(b) Schedule based on the available capacity in the pools - assuming the condition is `thinprovision=false`

May be a few more case can come up that affects scheduling. Can the code be designed to extend / configure which scheduler logic to apply.

Reply:
as the scheduler .. makes its passes, probably we can update the ZFSVolume CR with some events or status or conditions..

Reply:
that's correct @kmova, Right now we don't have any info related to pool, so we have no idea about the capacity available in the pool or any info related to Zpool. Later Once we have that info, we can consider that in our scheduling logic.

Reply:
can the csi node driver -- run some zpool commands and get the capcity of the pool, Is there any API like node capabilities that we can implement to enforce this limit on capacity.

Reply:
the volume create request will go to the csi controller only, so it can not know the available capacity by runing zpool command as it will be running on a one node only. We have to have a pool resources which can be used for scheduling decesion and we have to manage that resource also..